src/Spago/Build.hs

```diff
@@ -63,6 +65,22 @@ build BuildOptions{..} maybePostBuild = do
   let allPsGlobs = Packages.getGlobs   deps depsOnly configSourcePaths <> sourcePaths
       allJsGlobs = Packages.getJsGlobs deps depsOnly configSourcePaths <> sourcePaths
 
+      checkImports globs = do
+        graph <- Purs.graph globs
+        case graph of
+          Left err -> die [ displayShow err ]
+          Right (ModuleGraph moduleGraph) -> do
+            let
+              getPackage = fmap PackageName . headMay . drop 1 . dropWhile (/= ".spago") . Text.split (== '/')
+              spagoPackages = Map.fromList $ mapMaybe (\((moduleName, ModuleGraphNode{..})) -> (moduleName,) <$> getPackage path) $ Map.toList moduleGraph
+              sourceModuleDeps = Set.fromList (depends . snd =<< filter (isNothing . getPackage . path . snd) (Map.toList moduleGraph))
+              importedPackages = Set.fromList $ mapMaybe (flip Map.lookup spagoPackages) (Set.toList sourceModuleDeps)
+              importedTransitive = importedPackages `Set.difference` Set.fromList dependencies
```

Comment:
This is good, but we'll need a bit more logic for this to be correct.
First of all - as you might have noticed from the failing tests - we should classify the graph nodes in three kinds:
1. project modules: well, the modules in your project
2. direct dependency modules: nodes that are children of project modules but are not project modules themselves
3. transitive dependency modules: all the other nodes (note: this includes also nodes that might be in "direct dependency packages", but that are not directly imported in project files)

What we care about for this check is the set of (2). From that set of modules we can get a set of package names - let's call that `B`. If we call `A` the set of declared dependencies (the ones in your config), then:
- if a package is in `A` but not in `B`, then it's unused and we should emit a warning about it
- if a package is in `B` but not in `A`, then we're importing a transitive dependency and we should fail the build